src/dags/dag.py

```diff
@@ -29,30 +31,55 @@
     )
 
 
-@dataclass
+@dataclass(frozen=True)
 class FunctionExecutionInfo:
     """Information about a function that is needed to execute it.
 
     Attributes
     ----------
+        name: The name of the function.
         func: The function to execute.
-        argument_annotations: The argument annotations of the function.
-        return_annotation: The return annotation of the function.
 
     Properties
     ----------
+        annotations: The annotations of the function. For standard functions this
+            coincides with the __annotations__ attribute of the function. For partialled
+            functions, this is a dictionary with the names of the free arguments as keys
+            and their expected types as values, as well as the return type of the
+            function stored under the key "return". We assume that the type annotations
```

Comment:
```suggestion
            function stored under the key "return". Type annotations must be strings, else a `NonStringAnnotationError` is raised.
```

Maybe also add the gist of the `NonStringAnnotationError` message to the docstring so that people baffled by this sentence (I would have been three days ago)  know what to do before running into the error.